cluster/swarm/cluster.go

```diff
@@ -66,35 +66,16 @@ func (c *Cluster) Handle(e *cluster.Event) error {
 
 // CreateContainer aka schedule a brand new container into the cluster.
 func (c *Cluster) CreateContainer(config *dockerclient.ContainerConfig, name string) (*cluster.Container, error) {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
 
-	c.RLock()
-	defer c.RUnlock()
-retry:
-	// FIXME: to prevent a race, we check again after the pull if the node can still handle
-	// the container. We should store the state in the store before pulling and use this to check
-	// all the other container create, but, as we don't have a proper store yet, this temporary solution
-	// was chosen.
 	n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), config)
 	if err != nil {
 		return nil, err
 	}
 
 	if nn, ok := n.(*node); ok {
-		container, err := nn.create(config, name, false)
```

Comment:
Thanks @smothiki, great idea.

This is a temporary fix to get 0.2.0 out of the door since being race-condition free is a top priority, and being fast comes way after. I definitely agree we need to optimize this, but for now it does the work since only `docker run` and `docker rm` would be locked, all the other commands would remain lock free.

Maybe this is exactly what you just described, but I think a simple implementation for your solution would be to have `addtoqueue` (for instance) add the config to the internal `containers` map of Node with status `pending` or something.

That way, there is no need to explicitly handle it in the scheduler: `node.Containers()` will return both actual running containers and scheduled ones.

Maybe we should open an issue for this?